// @datadog/browser-rum/cjs/domain/record/record.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.record = void 0;
var tslib_1 = require("tslib");
var types_1 = require("../../types");
var serialize_1 = require("./serialize");
var observer_1 = require("./observer");
var types_2 = require("./types");
var utils_1 = require("./utils");
var mutationObserver_1 = require("./mutationObserver");
function record(options) {
    var emit = options.emit;
    // runtime checks for user options
    if (!emit) {
        throw new Error('emit function is required');
    }
    var mutationController = new mutationObserver_1.MutationController();
    var takeFullSnapshot = function () {
        var _a, _b, _c, _d;
        mutationController.flush(); // process any pending mutation before taking a full snapshot
        emit({
            data: {
                height: utils_1.getWindowHeight(),
                href: window.location.href,
                width: utils_1.getWindowWidth(),
            },
            type: types_1.RecordType.Meta,
        });
        emit({
            data: {
                has_focus: document.hasFocus(),
            },
            type: types_1.RecordType.Focus,
        });
        emit({
            data: {
                node: serialize_1.serializeDocument(document, options.initialPrivacyLevel),
                initialOffset: {
                    left: window.pageXOffset !== undefined
                        ? window.pageXOffset
                        : (document === null || document === void 0 ? void 0 : document.documentElement.scrollLeft) || ((_b = (_a = document === null || document === void 0 ? void 0 : document.body) === null || _a === void 0 ? void 0 : _a.parentElement) === null || _b === void 0 ? void 0 : _b.scrollLeft) || (document === null || document === void 0 ? void 0 : document.body.scrollLeft) ||
                            0,
                    top: window.pageYOffset !== undefined
                        ? window.pageYOffset
                        : (document === null || document === void 0 ? void 0 : document.documentElement.scrollTop) || ((_d = (_c = document === null || document === void 0 ? void 0 : document.body) === null || _c === void 0 ? void 0 : _c.parentElement) === null || _d === void 0 ? void 0 : _d.scrollTop) || (document === null || document === void 0 ? void 0 : document.body.scrollTop) ||
                            0,
                },
            },
            type: types_1.RecordType.FullSnapshot,
        });
    };
    takeFullSnapshot();
    var stopObservers = observer_1.initObservers({
        mutationController: mutationController,
        initialPrivacyLevel: options.initialPrivacyLevel,
        inputCb: function (v) {
            return emit({
                data: tslib_1.__assign({ source: types_2.IncrementalSource.Input }, v),
                type: types_1.RecordType.IncrementalSnapshot,
            });
        },
        mediaInteractionCb: function (p) {
            return emit({
                data: tslib_1.__assign({ source: types_2.IncrementalSource.MediaInteraction }, p),
                type: types_1.RecordType.IncrementalSnapshot,
            });
        },
        mouseInteractionCb: function (d) {
            return emit({
                data: tslib_1.__assign({ source: types_2.IncrementalSource.MouseInteraction }, d),
                type: types_1.RecordType.IncrementalSnapshot,
            });
        },
        mousemoveCb: function (positions, source) {
            return emit({
                data: {
                    positions: positions,
                    source: source,
                },
                type: types_1.RecordType.IncrementalSnapshot,
            });
        },
        mutationCb: function (m) {
            return emit({
                data: tslib_1.__assign({ source: types_2.IncrementalSource.Mutation }, m),
                type: types_1.RecordType.IncrementalSnapshot,
            });
        },
        scrollCb: function (p) {
            return emit({
                data: tslib_1.__assign({ source: types_2.IncrementalSource.Scroll }, p),
                type: types_1.RecordType.IncrementalSnapshot,
            });
        },
        styleSheetRuleCb: function (r) {
            return emit({
                data: tslib_1.__assign({ source: types_2.IncrementalSource.StyleSheetRule }, r),
                type: types_1.RecordType.IncrementalSnapshot,
            });
        },
        viewportResizeCb: function (d) {
            return emit({
                data: tslib_1.__assign({ source: types_2.IncrementalSource.ViewportResize }, d),
                type: types_1.RecordType.IncrementalSnapshot,
            });
        },
        focusCb: function (data) {
            return emit({
                type: types_1.RecordType.Focus,
                data: data,
            });
        },
    });
    return {
        stop: stopObservers,
        takeFullSnapshot: takeFullSnapshot,
        flushMutations: function () { return mutationController.flush(); },
    };
}
exports.record = record;
