// @datadog/browser-logs/cjs/domain/trackNetworkError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackNetworkError = void 0;
var browser_core_1 = require("@datadog/browser-core");
function trackNetworkError(configuration, errorObservable) {
    browser_core_1.startXhrProxy().onRequestComplete(function (context) { return handleCompleteRequest(browser_core_1.RequestType.XHR, context); });
    browser_core_1.startFetchProxy().onRequestComplete(function (context) { return handleCompleteRequest(browser_core_1.RequestType.FETCH, context); });
    function handleCompleteRequest(type, request) {
        if (!configuration.isIntakeUrl(request.url) && (isRejected(request) || isServerError(request))) {
            errorObservable.notify({
                message: format(type) + " error " + request.method + " " + request.url,
                resource: {
                    method: request.method,
                    statusCode: request.status,
                    url: request.url,
                },
                source: browser_core_1.ErrorSource.NETWORK,
                stack: truncateResponseText(request.responseText, configuration) || 'Failed to load',
                startClocks: request.startClocks,
            });
        }
    }
    return {
        stop: function () {
            browser_core_1.resetXhrProxy();
            browser_core_1.resetFetchProxy();
        },
    };
}
exports.trackNetworkError = trackNetworkError;
function isRejected(request) {
    return request.status === 0 && request.responseType !== 'opaque';
}
function isServerError(request) {
    return request.status >= 500;
}
function truncateResponseText(responseText, configuration) {
    if (responseText && responseText.length > configuration.requestErrorResponseLengthLimit) {
        return responseText.substring(0, configuration.requestErrorResponseLengthLimit) + "...";
    }
    return responseText;
}
function format(type) {
    if (browser_core_1.RequestType.XHR === type) {
        return 'XHR';
    }
    return 'Fetch';
}
